src/test/java/fi/fmi/avi/converter/tac/bulletin/GenericAviationWeatherMessageParserTest.java

```diff
@@ -0,0 +1,192 @@
+package fi.fmi.avi.converter.tac.bulletin;
+
+import static junit.framework.TestCase.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+import java.time.ZonedDateTime;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+import org.springframework.test.context.support.AnnotationConfigContextLoader;
+
+import fi.fmi.avi.converter.AviMessageConverter;
+import fi.fmi.avi.converter.ConversionHints;
+import fi.fmi.avi.converter.ConversionResult;
+import fi.fmi.avi.converter.tac.TACTestConfiguration;
+import fi.fmi.avi.converter.tac.conf.TACConverter;
+import fi.fmi.avi.converter.tac.lexer.AviMessageLexer;
+import fi.fmi.avi.model.GenericAviationWeatherMessage;
+import fi.fmi.avi.model.MessageType;
+import fi.fmi.avi.model.PartialDateTime;
+import fi.fmi.avi.model.PartialOrCompleteTimeInstant;
+import fi.fmi.avi.model.PartialOrCompleteTimePeriod;
+
+@RunWith(SpringJUnit4ClassRunner.class)
+@ContextConfiguration(classes = TACTestConfiguration.class, loader = AnnotationConfigContextLoader.class)
+public class GenericAviationWeatherMessageParserTest {
+    @Autowired
+    private AviMessageConverter converter;
+
+    @Autowired
+    private AviMessageLexer lexer;
+
+    @Test
+    public void tafTest() {
+        String tacMessage = "TAF EVRA 301103Z 3012/3112 15013KT 8000 OVC008\r\n" //
+                + "TEMPO 3012/3015 14015G26KT 5000 -RA OVC010\r\n" //
+                + "FM301500 18012KT 9000 NSW SCT015 BKN020\r\n" //
+                + "TEMPO 3017/3103 19020G33KT 3000 -SHRA BKN012CB BKN020=";
+        ConversionResult<GenericAviationWeatherMessage> result = converter.convertMessage(tacMessage, TACConverter.TAC_TO_GENERIC_AVIATION_WEATHER_MESSAGE_POJO,
+                new ConversionHints());
+
+        assertEquals(ConversionResult.Status.SUCCESS, result.getStatus());
+
+        GenericAviationWeatherMessage msg = result.getConvertedMessage().get();
+
+        assertEquals(tacMessage, msg.getOriginalMessage());
+        assertEquals(MessageType.TAF, msg.getMessageType().get());
+        assertEquals(GenericAviationWeatherMessage.Format.TAC, msg.getMessageFormat());
+        assertEquals("--30T11:03Z",
+                msg.getIssueTime().map(PartialOrCompleteTimeInstant::getPartialTime).map(Optional::get).map(PartialDateTime::toString).orElse(null));
+        assertEquals("--30T12:Z", msg.getValidityTime()
+                .map(PartialOrCompleteTimePeriod::getStartTime)
+                .map(Optional::get)
```

Comment:
Hint: `map(PartialOrCompleteTimeInstant::getStartTime).map(Optional::get)` and similar cases can be written shorter and clearer `flatMap(PartialOrCompleteTimeInstant::getStartTime)`. I would suggest to do a global search for `map(Optional::get)`.